File: ff4j-spring-boot-starter/src/main/java/org/ff4j/spring/boot/domain/FeatureApiBean.java
package org.ff4j.spring.boot.domain;


import lombok.Getter;
import lombok.Setter;
import org.ff4j.core.Feature;
import org.ff4j.property.Property;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Paul
 *
 * @author <a href="mailto:paul58914080@gmail.com">Paul Williams</a>
 */
public class FeatureApiBean implements Serializable {

    private static final long serialVersionUID = -4977143873952901044L;

    @Getter
    @Setter
    private String uid;

    @Getter
    @Setter
    private boolean enable;

    @Getter
    @Setter
    private String description;

    @Getter
    @Setter
    private String group;

    @Getter
    @Setter
    private List<String> permissions = new ArrayList<>();

    @Getter
    @Setter
    private FlippingStrategyApiBean flippingStrategy;

    @Getter
    @Setter
    private Map<String, PropertyApiBean> customProperties = new HashMap<>();

    public FeatureApiBean() {
        super();
    }

    public FeatureApiBean(Feature f) {
        this.uid = f.getUid();
        this.enable = f.isEnable();
        this.description = f.getDescription();
        this.group = f.getGroup();
        this.permissions = new ArrayList<>(f.getPermissions());
        if (f.getFlippingStrategy() != null) {
            this.flippingStrategy = new FlippingStrategyApiBean(f.getFlippingStrategy());
        }
        if (f.getCustomProperties() != null) {
            for (Property<?> ap1 : f.getCustomProperties().values()) {
                customProperties.put(ap1.getName(), new PropertyApiBean(ap1));
            }
        }
    }
}

File: ff4j-spring-services/src/main/kotlin/org/ff4j/services/domain/FeatureApiBean.kt
/*-
 * #%L
 * ff4j-spring-services
 * %%
 * Copyright (C) 2013 - 2023 FF4J
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package org.ff4j.services.domain

import org.ff4j.core.Feature
import org.ff4j.exception.FeatureAccessException
import org.ff4j.services.exceptions.FlippingStrategyBadRequestException
import org.ff4j.services.exceptions.PropertiesBadRequestException
import org.ff4j.utils.MappingUtil
import java.io.Serializable

/**
 * @author [Paul Williams](mailto:paul58914080@gmail.com)
 */
class FeatureApiBean : Serializable {

  companion object {
    private const val serialVersionUID = -4977143873952901044L
  }

  var uid: String? = null
  var description: String? = null
  var group: String? = null
  var flippingStrategy: FlippingStrategyApiBean? = null
  var permissions: MutableSet<String> = HashSet()
  var customProperties: MutableMap<String, PropertyApiBean> = HashMap()
  var enable: Boolean = false

  constructor() : super()

  constructor(feature: Feature) {
    this.uid = feature.uid
    this.enable = feature.isEnable
    this.description = feature.description
    this.permissions = feature.permissions
    this.group = feature.group
    feature.flippingStrategy?.let {
      this.flippingStrategy = FlippingStrategyApiBean(it)
    }
    feature.customProperties?.let {
      it.values.forEach { property ->
        run {
          this.customProperties[property.name] = PropertyApiBean(property)
        }
      }
    }
  }

  fun toFeature(): Feature {
    val feature = Feature(uid)
    feature.description = description
    feature.isEnable = enable
    feature.group = group
    feature.permissions = permissions
    initFlippingStrategy(feature)
    initProperties(feature)
    return feature
  }

  private fun initFlippingStrategy(feature: Feature) {
    flippingStrategy?.let {
      try {
        feature.flippingStrategy = MappingUtil.instanceFlippingStrategy(uid, it.type, it.initParams)
      } catch (exception: FeatureAccessException) {
        throw FlippingStrategyBadRequestException(exception)
      }
    }
  }

  private fun initProperties(feature: Feature) {
    addProperty(customProperties.values, feature)
  }

  private fun addProperty(values: MutableCollection<PropertyApiBean>, feature: Feature) {
    values.forEach { propertyApiBean ->
      run {
        try {
          feature.addProperty(propertyApiBean.asProperty())
        } catch (exception: IllegalArgumentException) {
          throw PropertiesBadRequestException(exception)
        }
      }
    }
  }
}
